pkg/test/utils/kubernetes.go

```diff
@@ -536,3 +538,37 @@ func WaitForCRDs(dClient discovery.DiscoveryInterface, crds []runtime.Object) er
 		return true, nil
 	})
 }
+
+// TestEnvironment is a struct containing the envtest environment, Kubernetes config and clients.
+type TestEnvironment struct {
+	Environment     *envtest.Environment
+	Config          *rest.Config
+	Client          client.Client
+	DiscoveryClient discovery.DiscoveryInterface
+}
+
+// StartTestEnvironment is a wrapper for controller-runtime's envtest that creates a Kubernetes API server and etcd
+// suitable for use in tests.
+func StartTestEnvironment() (env TestEnvironment, err error) {
+	env.Environment = &envtest.Environment{}
+
+	// Retry up to three times for the test environment to start up in case there is a port collision (#510).
+	for i := 0; i < 3; i++ {
+		env.Config, err = env.Environment.Start()
+		if err == nil {
+			break
+		}
+	}
+
+	if err != nil {
+		return
+	}
+
+	env.Client, err = client.New(env.Config, client.Options{})
+	if err != nil {
+		return
+	}
+
+	env.DiscoveryClient, err = discovery.NewDiscoveryClientForConfig(env.Config)
```

Comment:
we are already in the `discovery` package...  could we rename `NewDiscoveryClientForConfig` to `New` or `NewClient`... it currently seems repetitive and redundant :)

Reply:
It's a Kubernetes method :)

https://godoc.org/github.com/kubernetes/client-go/discovery#NewDiscoveryClientForConfig

Reply:
Also this is the utils package.

Reply:
haha... 🇺🇬